applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp

```diff
@@ -371,187 +340,260 @@ namespace Kratos
+                                        const GeoMechanicsNewtonRaphsonErosionProcessStrategyType::Pointer& p_solving_strategy,
+                                        const CallBackFunctions& rCallBackFunctions)
+    {
 
-                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Finished writing result" << std::endl;
+        auto currentProcess = std::static_pointer_cast<GeoFlowApplyConstantHydrostaticPressureProcess>(RiverBoundary);
+        KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "River boundary name: " << currentProcess->GetName() << std::endl;
+
+        currentProcess->SetReferenceCoord(criticalHeadInfo.minCriticalHead);
+        mCurrentHead = criticalHeadInfo.minCriticalHead;
+        mCriticalHead = mCurrentHead;
+
+        std::vector<Element *> pipeElements;
+        pipeElements = p_solving_strategy->GetPipingElements();
+        const auto noPipeElements = pipeElements.size();
+
+        int step = 1;
+        const auto maxSteps = static_cast<int>(std::ceil((criticalHeadInfo.maxCriticalHead - criticalHeadInfo.minCriticalHead) / criticalHeadInfo.stepCriticalHead)) + 2;
+
+
+        while (!mExitLoop)
+        {
+            if (criticalHeadInfo.maxCriticalHead - mCriticalHead < -1e-9)
+            {
+                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Critical head undetermined at " << mCriticalHead << ", max search head reached: " << criticalHeadInfo.maxCriticalHead << std::endl;
+                mExitLoop = true;
+                break;
             }
 
-            rLogCallback(kratosLogBuffer.str().c_str());
-            Logger::RemoveOutput(p_output);
+            KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Searching at head: " << mCurrentHead << std::endl;
 
-            ResetModelParts();
-            return 0;
+            std::ostringstream currentHeadStream;
+            currentHeadStream << std::setprecision(8) << std::noshowpoint << mCurrentHead;
+            std::string currentHeadString = currentHeadStream.str();
+
+            std::string progress = "Calculating head level " + currentHeadString + "m (" + std::to_string(step) + "/" + std::to_string(maxSteps) + ")";
+            rCallBackFunctions.rReportTextualProgress(progress.data());
+            rCallBackFunctions.rReportProgress(((double)step) / ((double)maxSteps));
+
+            MainExecution(model_part, p_solving_strategy, 0.0, 1.0, 1);
+
+            auto count = std::size_t{0};
+            for (Element *element : pipeElements)
+            {
+                if (element->GetValue(PIPE_ACTIVE))
+                    count += 1;
+            }
+
+            KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Open pipe elements: " << count << std::endl;
+
+            if (count == noPipeElements)
+            {
+                HandleCriticalHeadFound(criticalHeadInfo);
+                mExitLoop = true;
```

Comment:
There used to be a `break` here, which means that we would jump out of the loop. Now we continue until the end of loop body. I'm not sure whether that was your intention, because the behavior is different now.